src/Magento/FunctionalTestingFramework/StaticCheck/TestDependencyCheck.php

```diff
@@ -0,0 +1,385 @@
+                    $parameterizedReference,
+                    $arguments
+                );
+                $splitArguments = explode(',', ltrim(rtrim($arguments[0], ")"), "("));
+                foreach ($splitArguments as $argument) {
+                    // Do nothing for 'string' or $persisted.data$
+                    if (preg_match(ActionObject::STRING_PARAMETER_REGEX, $argument)) {
+                        continue;
+                    } elseif (preg_match(TestGenerator::PERSISTED_OBJECT_NOTATION_REGEX, $argument)) {
+                        continue;
+                    }
+                    // trim `data.field` to `data`
+                    preg_match('/([^.]+)/', $argument, $entityName);
+                    // Double check that {{data.field}} isn't an argument for an ActionGroup
+                    $entity = $this->findEntity($entityName[1]);
+                    preg_match_all(self::ACTIONGROUP_ARGUMENT_REGEX_PATTERN, $contents, $possibleArgument);
+                    if (array_search($entityName[1], $possibleArgument[1]) !== false) {
+                        continue;
+                    }
+                    if ($entity !== null) {
+                        $allEntities[$entity->getName()] = $entity;
+                    }
+                }
+            }
+            // Check actionGroup references
+            foreach ($actionGroupReferences[1] as $reference) {
+                $entity = $this->findEntity($reference);
+                if ($entity !== null) {
+                    $allEntities[$entity->getName()] = $entity;
+                }
+            }
+            // Check extended objects
+            foreach ($extendReferences[1] as $reference) {
+                $entity = $this->findEntity($reference);
+                if ($entity !== null) {
+                    $allEntities[$entity->getName()] = $entity;
+                }
+            }
+
+            $currentModule = $this->moduleNameToComposerName[$moduleFullName];
+            $modulesReferencedInTest = $this->getModuleDependenciesFromReferences($allEntities, $currentModule);
+            $moduleDependencies = $this->flattenedDependencies[$moduleFullName];
+            // Find Violations
+            $violatingReferences = [];
+            foreach ($modulesReferencedInTest as $entityName => $files) {
+                $valid = false;
+                foreach ($files as $module) {
+                    if (array_key_exists($module, $moduleDependencies) || $module == $currentModule) {
+                        $valid = true;
+                        break;
+                    }
+                }
+                if (!$valid) {
+                    $violatingReferences[$entityName] = $files;
+                }
+            }
+
+            if (!empty($violatingReferences)) {
+                // Build error output
+                $errorOutput = "\nFile \"{$filePath->getRealPath()}\"\ncontains references to following modules:\n\t\t";
```

Comment:
Should we point out the violation directly? e.g. using words like "SomeTest should not contain references to SomeModule" would be better?

Reply:
I can change the error output to be more precise, yes.